### PR TITLE
[Model] Add MiDashengLM model support

### PR DESCRIPTION
**Pull request by bingchen-mi**
This is for support MiDashengLM model.

The model and technical report are open-sourced in https://huggingface.co/mispeech/midashenglm-7b

Thanks for your review!

All the best!

### CODE REVIEW

File: vllm/model_executor/models/midashenglm.py [hunk 1]
```diff
@@ -0,0 +1,797 @@
...
+            quant_config=quant_config,
+            prefix=f"{prefix}.qkv",
+        )
+        self.attn = MultiHeadAttention(
+            self.num_heads,
+            self.head_dim,
+            self.scale,
+            num_kv_heads=self.num_kv_heads,
+        )
+        self.attn_drop = nn.Dropout(attn_drop)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/midashenglm.py]: No need to include dropout if it's not intended for inference

File: vllm/model_executor/models/midashenglm.py [hunk 2]
```diff
@@ -0,0 +1,797 @@
...
+        config: DashengConfig,
+        quant_config: Optional[QuantizationConfig] = None,
+        prefix: str = "",
+    ):
+        super().__init__(config)
+
+        self.target_length = config.target_length
+        self.embed_dim = config.embed_dim
+        self.hop_length = config.hop_length
+        self.gradient_checkpointing = False
```
Comment (DarkLight1337) [inline vllm/model_executor/models/midashenglm.py]: No need to inherit from `PreTrainedModel` unless there is some feature from it that is needed for inference

File: vllm/transformers_utils/processors/midashenglm.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+import numpy as np
+import torch
+from transformers import (AutoProcessor, Qwen2Tokenizer, Qwen2TokenizerFast,
+                          Wav2Vec2FeatureExtractor)
+from transformers.feature_extraction_utils import BatchFeature
+from transformers.processing_utils import (ProcessingKwargs, ProcessorMixin,
+                                           Unpack)
+
+
+class MiDashengLMProcessorKwargs(ProcessingKwargs):
```
Comment (DarkLight1337) [inline vllm/transformers_utils/processors/midashenglm.py]: If there is nothing changed from the processor definition from HF Hub, no need to implement it in vLLM. You can skip type checking inside the modeling file when getting HF processor

**Comment by DarkLight1337** [comment]
Thanks for implementing this model in vLLM, I have made some initial comments

File: vllm/model_executor/models/midashenglm.py (hunk 1 above)
Comment (bingchen-mi) [reply vllm/model_executor/models/midashenglm.py]: Deleted all Dropout.
thanks

File: vllm/model_executor/models/midashenglm.py (hunk 2 above)
Comment (bingchen-mi) [reply vllm/model_executor/models/midashenglm.py]: yes, has removed it.

File: vllm/transformers_utils/processors/midashenglm.py (hunk 3 above)
Comment (bingchen-mi) [reply vllm/transformers_utils/processors/midashenglm.py]: ok, the processor has nothing changed, remove it, and get it from HF Hub.

File: vllm/transformers_utils/processors/midashenglm.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+            tokenizer_init_kwargs=self.tokenizer.init_kwargs,
+            **kwargs,
+        )
+
+        if audio is not None:
+            audio = [self._validate_audio_sample(sample) for sample in audio]
+            # ensure we have as much audios as audio tokens
+            num_audio_tokens = sum(
+                sample.count(self.audio_token) for sample in text)
+            num_audios = 1 if type(audio) is np.ndarray else len(audio)
```
[the remark this answers is not in the document]
Comment (bingchen-mi) [reply vllm/transformers_utils/processors/midashenglm.py]: removed implement of processor from vllm

File: vllm/model_executor/models/midashenglm.py [hunk 5]
```diff
@@ -0,0 +1,784 @@
...
+        mask: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        x = x + self.ls1(self.attn(self.norm1(x), mask))
+        x = x + self.ls2(self.mlp(self.norm2(x)))
+        return x
+
+
+class DashengAudioTransformer(nn.Module):
+    config_class = DashengConfig
+    supports_gradient_checkpointing = True
```
Comment (DarkLight1337) [inline vllm/model_executor/models/midashenglm.py]: These fields are also not needed

**Review by DarkLight1337 — COMMENTED** [review]
Can you also add basic tests to this model, e.g. in `tests/models/multimodal/processing/test_common.py`?

File: vllm/model_executor/models/midashenglm.py (hunk 5 above)
Comment (bingchen-mi) [reply vllm/model_executor/models/midashenglm.py]: Remove it. thanks

**Comment by bingchen-mi** [comment]
> Can you also add basic tests to this model, e.g. in `tests/models/multimodal/processing/test_common.py`?

just need to add model_id? Do I need to write other test code?

**Comment by DarkLight1337** [comment]
You just have to add the model ID

File: tests/models/multimodal/processing/test_common.py [hunk 6]
```diff
@@ -317,6 +317,7 @@ def _test_processing_correctness_one(
     "google/paligemma2-3b-ft-docci-448",
     "microsoft/Phi-3.5-vision-instruct",
     "microsoft/Phi-4-multimodal-instruct",
+    "mispeech/midashenglm-7b",
```
Comment (DarkLight1337) [inline tests/models/multimodal/processing/test_common.py]: Please keep this in alphabetical order with respect to the model architecture (not the organization name)

Comment (bingchen-mi) [reply tests/models/multimodal/processing/test_common.py]: Got it, thanks

File: vllm/transformers_utils/configs/midashenglm.py
Comment (Isotr0py) [inline vllm/transformers_utils/configs/midashenglm.py:1]: If `trust_remote_code=True` can work well, there is no need to port this file.

File: vllm/model_executor/models/midashenglm.py [hunk 7]
```diff
@@ -0,0 +1,785 @@
...
+                init_values=config.init_values,
+                quant_config=quant_config,
+                prefix=f"{prefix}.block{i}",
+            ) for i in range(config.depth))
+        self.norm = nn.LayerNorm(config.embed_dim, eps=1e-6)
+
+    def _init_front_end(self, config):
+        ori_default_dtype = torch.get_default_dtype()
+        torch.set_default_dtype(torch.float32)
+
```
Comment (Isotr0py) [inline vllm/model_executor/models/midashenglm.py]: We can use `with set_default_torch_dtype(torch.float32):` context here.

File: vllm/model_executor/models/midashenglm.py [hunk 8]
```diff
@@ -0,0 +1,785 @@
...
+                output_size=out_dim,
+                quant_config=quant_config,
+                prefix=f"{prefix}.fc1",
+            ), get_act_fn("gelu"),
+            RowParallelLinear(
+                input_size=out_dim,
+                output_size=out_dim,
+                quant_config=quant_config,
+                prefix=f"{prefix}.fc2",
+            ))
```
Comment (Isotr0py) [inline vllm/model_executor/models/midashenglm.py]:
```suggestion
        self.net = nn.Sequential(
            ColumnParallelLinear(
                input_size=in_dim * self.k,
                output_size=out_dim,
                quant_config=quant_config,
                prefix=f"{prefix}.fc1",
                return_bias=False,
            ), get_act_fn("gelu"),
            RowParallelLinear(
                input_size=out_dim,
                output_size=out_dim,
                quant_config=quant_config,
                prefix=f"{prefix}.fc2",
                return_bias=False,
            ))
```
You can add `return_bias=False` to let parallel linear returns like normal `nn.Linear`

File: vllm/model_executor/models/midashenglm.py [hunk 9]
```diff
@@ -0,0 +1,785 @@
...
+            if mask is not None:
+                mask = mask[:, :-num_frames_to_discard]
+        if mask is None:
+            mask = torch.ones(x.shape[:-1], dtype=torch.long, device=x.device)
+        x = x.reshape(batch_size, -1, self.k *
+                      dim)  # rearrange(x, "b (s k) d -> b s (k d)", k=self.k)
+        for layer in self.net:
+            x = layer(x)
+            if isinstance(x, tuple):
+                x, _ = x
```
Comment (Isotr0py) [inline vllm/model_executor/models/midashenglm.py]:
```suggestion
```

File: vllm/model_executor/models/midashenglm.py [hunk 10]
```diff
@@ -0,0 +1,785 @@
...
+        sampling_metadata: SamplingMetadata,
+    ) -> Optional[torch.Tensor]:
+        return self.decoder.compute_logits(hidden_states, sampling_metadata)
+
+    def load_weights(self, weights: Iterable[tuple[str,
+                                                   torch.Tensor]]) -> set[str]:
+        loader = AutoWeightsLoader(
+            self,
+            skip_prefixes=["audio_encoder.front_end."],
+        )
```
Comment (Isotr0py) [inline vllm/model_executor/models/midashenglm.py]:
```suggestion
        loader = AutoWeightsLoader(self)
```
Seems the model doesn't have parameters in `audio_encoder.front_end`?

File: vllm/transformers_utils/configs/midashenglm.py
Comment (bingchen-mi) [reply vllm/transformers_utils/configs/midashenglm.py:1]: I need to set `self.text_config.rope_scaling = None` in config, Is there somewhere else I can set it up?

Comment (Isotr0py) [reply vllm/transformers_utils/configs/midashenglm.py:1]: Ah, I see. Let's keep it like this then.

File: vllm/model_executor/models/midashenglm.py (hunk 7 above)
Comment (bingchen-mi) [reply vllm/model_executor/models/midashenglm.py]: Got it, thanks

File: vllm/model_executor/models/midashenglm.py (hunk 8 above)
Comment (bingchen-mi) [reply vllm/model_executor/models/midashenglm.py]: ok, thanks

File: vllm/model_executor/models/midashenglm.py (hunk 9 above)
Comment (bingchen-mi) [reply vllm/model_executor/models/midashenglm.py]: remove it.

File: vllm/model_executor/models/midashenglm.py (hunk 10 above)
Comment (bingchen-mi) [reply vllm/model_executor/models/midashenglm.py]: I found a little issue with weight loading when developing on v0.10.0, so I skipped that and it seems to be working fine now.
Removed the skip_prefixes, thanks

File: vllm/model_executor/models/midashenglm.py [hunk 11]
```diff
@@ -0,0 +1,780 @@
...
+        mask: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        t = x.shape[-1]
+        x = x + self.time_pos_embed[:, :, :, :t]
+        x = (x + self.freq_pos_embed[:, :, :, :]
+             )  # Just to support __getitem__ in posembed
+        x = torch.permute(torch.flatten(x, 2, 3),
+                          (0, 2, 1))  # rearrange(x, "b c f t -> b (f t) c")
+        for block in self.blocks:
+            if self.gradient_checkpointing and self.training:
```
Comment (jeejeelee) [inline vllm/model_executor/models/midashenglm.py]: We need to remove this unused code.

File: vllm/model_executor/models/midashenglm.py [hunk 12]
```diff
@@ -0,0 +1,780 @@
...
+        config: DashengConfig,
+        quant_config: Optional[QuantizationConfig] = None,
+        prefix: str = "",
+    ):
+        super().__init__()
+
+        self.target_length = config.target_length
+        self.embed_dim = config.embed_dim
+        self.hop_length = config.hop_length
+        self.gradient_checkpointing = False
```
Comment (jeejeelee) [inline vllm/model_executor/models/midashenglm.py]: We need to remove this unused var

File: vllm/model_executor/models/midashenglm.py [hunk 13]
```diff
@@ -0,0 +1,780 @@
...
+        prefix: str = "",
+    ):
+        super().__init__()
+        self.k = downsample_rate
+        self.net = nn.Sequential(
+            ColumnParallelLinear(
+                input_size=in_dim * self.k,
+                output_size=out_dim,
+                quant_config=quant_config,
+                prefix=f"{prefix}.fc1",
```
Comment (jeejeelee) [inline vllm/model_executor/models/midashenglm.py]: This should be `net.0`, right?
```suggestion
                prefix=f"{prefix}.net.0",
```

File: vllm/model_executor/models/midashenglm.py [hunk 14]
```diff
@@ -0,0 +1,780 @@
...
+                return_bias=False,
+            ), get_act_fn("gelu"),
+            RowParallelLinear(
+                input_size=out_dim,
+                output_size=out_dim,
+                quant_config=quant_config,
+                prefix=f"{prefix}.fc2",
```
Comment (jeejeelee) [inline vllm/model_executor/models/midashenglm.py]: ditto
```suggestion
                prefix=f"{prefix}.net.1",
```

File: vllm/model_executor/models/midashenglm.py (hunk 11 above)
Comment (bingchen-mi) [reply vllm/model_executor/models/midashenglm.py]: removed `gradient_checkpointing `

File: vllm/model_executor/models/midashenglm.py (hunk 12 above)
Comment (bingchen-mi) [reply vllm/model_executor/models/midashenglm.py]: removed unused var, thanks

File: vllm/model_executor/models/midashenglm.py (hunk 13 above)
Comment (bingchen-mi) [reply vllm/model_executor/models/midashenglm.py]: yes, thanks

File: vllm/model_executor/models/midashenglm.py (hunk 14 above)
Comment (bingchen-mi) [reply vllm/model_executor/models/midashenglm.py]: thanks

Comment (jeejeelee) [reply vllm/model_executor/models/midashenglm.py]: Apologies, the prefix should be modified to `net.2`. Refer to: https://huggingface.co/mispeech/midashenglm-7b/blob/main/model.safetensors.index.json#L403

File: vllm/model_executor/models/midashenglm.py [hunk 15]
```diff
@@ -0,0 +1,775 @@
...
+            flatten=False,
+            patch_stride=config.patch_stride,
+        )
+
+        self.time_pos_embed = nn.Parameter(
+            torch.randn(1, config.embed_dim, 1, self.patch_embed.grid_size[1])
+            * 0.02)
+        self.freq_pos_embed = nn.Parameter(
+            torch.randn(1, config.embed_dim, self.patch_embed.grid_size[0], 1)
+            * 0.02)
```
Comment (Isotr0py) [inline vllm/model_executor/models/midashenglm.py]:
```suggestion
        self.time_pos_embed = nn.Parameter(
            torch.empty(1, config.embed_dim, 1, self.patch_embed.grid_size[1]))
        self.freq_pos_embed = nn.Parameter(
            torch.empty(1, config.embed_dim, self.patch_embed.grid_size[0], 1))
```
Let's use `torch.empty` to avoid extra calculation, because initial parameter distribution doesn't matter for inference.

**Review by Isotr0py — APPROVED** [review]
LGTM! Just leave a nit.

File: vllm/model_executor/models/midashenglm.py (hunk 15 above)
Comment (bingchen-mi) [reply vllm/model_executor/models/midashenglm.py]: learn it, thanks

File: vllm/model_executor/models/midashenglm.py [hunk 16]
```diff
@@ -0,0 +1,773 @@
...
+                output_size=out_dim,
+                quant_config=quant_config,
+                prefix=f"{prefix}.net.0",
+                return_bias=False,
+            ), get_act_fn("gelu"),
+            RowParallelLinear(
+                input_size=out_dim,
+                output_size=out_dim,
+                quant_config=quant_config,
+                prefix=f"{prefix}.net.1",
```
Comment (jeejeelee) [inline vllm/model_executor/models/midashenglm.py]: Can you confirm whether this is `net.1` or `net.fc.2`?

Comment (bingchen-mi) [reply vllm/model_executor/models/midashenglm.py]: The source code of the model is as follows:
```
class AudioProjectorSubsample(nn.Module):
    def __init__(
        self,
        in_dim: int,
        out_dim: int,
        downsample_rate=5,
        dtype: Optional[torch.dtype] = None,
    ):
        super().__init__()
        self.k = downsample_rate
        self.net = nn.Sequential(
            nn.Linear(in_dim * self.k, out_dim, dtype=dtype),
            nn.GELU(),
            nn.Linear(out_dim, out_dim, dtype=dtype),
        )
```
maybe modified to `net.2`  ?

Comment (jeejeelee) [reply vllm/model_executor/models/midashenglm.py]: Yes, you are right

Comment (bingchen-mi) [reply vllm/model_executor/models/midashenglm.py]: ok, thanks

**Review by jeejeelee — APPROVED** [review]
overall, LGTM

**Comment by bingchen-mi** [comment]
There's an issue discovered in the previous CI run: `audio_len` is too short, causing a ValueError to be raised.
According to `calculate_mel_frames_dasheng`, when `audio_length_samples < 3040` result is 0.
the midashenglm model requires a minimum of 200ms of audio (at a 16k sampling rate) 
```
            if num_features == 0:
                audios = mm_items.get_items("audio", AudioProcessorItems)
                audio_len = audios.get_audio_length(item_idx)

                raise ValueError(f"The audio (len={audio_len}) is too short "
                                 "to be represented inside the model")
```
I'm also curious why qwen2_audio doesn't trigger this error during testing.

**Comment by Isotr0py** [comment]
IIRC, Qwen2-Audio will do padding during audio processing to make sure to output at least one audio frame: https://github.com/huggingface/transformers/blob/e690fe61e82e1b83c4b30a86559cf9c9d1bb0356/src/transformers/models/qwen2_audio/processing_qwen2_audio.py#L124-L127

**Comment by bingchen-mi** [comment]
> IIRC, Qwen2-Audio will do padding during audio processing to make sure to output at least one audio frame: https://github.com/huggingface/transformers/blob/e690fe61e82e1b83c4b30a86559cf9c9d1bb0356/src/transformers/models/qwen2_audio/processing_qwen2_audio.py#L124-L127

thanks!

**Comment by bingchen-mi** [comment]
I update some code to pass the test_common test.
Does the current CI is related to this PR?

**Comment by DarkLight1337** [comment]
The CI failure is not related - merging
